dialogue-java-client/src/main/java/com/palantir/dialogue/UrlBuilder.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.CharMatcher;
+import com.google.common.base.Joiner;
+import com.google.common.collect.Iterables;
+import com.google.common.net.InetAddresses;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.UnsafeArg;
+import java.io.ByteArrayOutputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+/** A simplistic URL builder, not tuned for performance. */
+public final class UrlBuilder {
+
+    private static final Joiner SLASH_JOINER = Joiner.on('/');
+
+    private final String protocol;
+    private String host;
+    private int port = -1;
+    private List<String> pathSegments = new ArrayList<>();
+    private List<String> queryNamesAndValues = new ArrayList<>(); // alternating (name, value) pairs
```

Comment:
I think it may be clearer to use a Multimap

Reply:
sorry, thought I had commented on your previous comment; I used a list to save some object allocations. maybe it doesn't matter to much. switching to multimap.

Reply:
ok